include/deal.II/trilinos/nox.h

```diff
@@ -0,0 +1,250 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2022 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_trilinos_nox
+#define dealii_trilinos_nox
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_TRILINOS
+
+#  include <deal.II/base/exceptions.h>
+
+#  include <deal.II/lac/solver_control.h>
+
+#  include <Teuchos_ParameterList.hpp>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace TrilinosWrappers
+{
+  // Indicate that NOXSolver has not converged.
+  DeclException0(ExcNOXNoConvergence);
+
+
+  /**
+   * Wrapper around the non-linear solver from the NOX
+   * packge (https://docs.trilinos.org/dev/packages/nox/doc/html/index.html),
+   * targeting deal.II data structures.
+   */
+  template <typename VectorType>
+  class NOXSolver
+  {
+  public:
+    /**
+     * Struct that helps to configure NOXSolver. More advanced
+     * parameters are passed to the constructor NOXSolver
+     * directly via a Teuchos::ParameterList.
+     */
+    struct AdditionalData
+    {
+    public:
+      /**
+       * Constructor.
+       */
+      AdditionalData(const unsigned int max_iter                       = 10,
+                     const double       abs_tol                        = 1.e-20,
+                     const double       rel_tol                        = 1.e-5,
+                     const unsigned int threshold_nonlinear_iterations = 1,
+                     const unsigned int threshold_n_linear_iterations  = 0,
+                     const bool         reuse_solver                   = false);
+
+      /**
+       * Max number of non-linear iterations.
+       */
+      unsigned int max_iter;
+
+      /**
+       * Absolute l2 tolerance to be reached.
+       */
+      double abs_tol;
+
+      /**
+       * Relative l2 tolerance to be reached.
+       */
+      double rel_tol;
+
+      /**
+       * Number of non-linear iterations after which the preconditioner
+       * should be updated.
+       */
+      unsigned int threshold_nonlinear_iterations;
+
+      /**
+       * Max number of linear iterations after which the preconditioner
+       * should be updated. This is only used if a lambda is attached to
```

Comment:
`std::function`s can be bound to free functions and class member functions as well. So maybe we should state that "This is only used if solve_with_jacobian_and_track_n_linear_iterations has been defined.", or something like that?

Reply:
To be dealt with in https://github.com/dealii/dealii/pull/14629